File: apps/backend/src/api/routes/activities.py
"""Activities API routes for CRM.

Provides full CRUD operations for managing activities/interactions tracking.
"""

from datetime import UTC, datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import joinedload

from src.cache.decorators import cached, invalidate_cache
from src.config.database import get_db
from src.db.crm_models import Activity as ActivityModel
from src.db.crm_models import Contact as ContactModel
from src.db.crm_schemas import (
    Activity,
    ActivityComplete,
    ActivityCreate,
    ActivityType,
    ActivityUpdate,
    ActivityWithRelations,
    PaginatedActivities,
)
from src.db.demo_models import Customer as CustomerModel
from src.db.demo_models import Order as OrderModel
from src.db.demo_models import Quote as QuoteModel
from src.services.sse_service import sse_service

router = APIRouter(prefix="/api/activities", tags=["activities"])


@router.get("", response_model=PaginatedActivities)
@cached(ttl=60, key_prefix="api_activities_list")  # 1 minute cache
async def list_activities(
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100),
    activity_type: ActivityType | None = None,
    customer_id: UUID | None = None,
    contact_id: UUID | None = None,
    order_id: UUID | None = None,
    quote_id: UUID | None = None,
    include_completed: bool = True,
    db: AsyncSession = Depends(get_db),
):
    """List activities with pagination and filters."""
    query = select(ActivityModel)

    # Apply filters
    if activity_type:
        query = query.where(ActivityModel.activity_type == activity_type.value)

    if customer_id:
        query = query.where(ActivityModel.customer_id == customer_id)

    if contact_id:
        query = query.where(ActivityModel.contact_id == contact_id)

    if order_id:
        query = query.where(ActivityModel.order_id == order_id)

    if quote_id:
        query = query.where(ActivityModel.quote_id == quote_id)

    if not include_completed:
        query = query.where(ActivityModel.completed_at.is_(None))

    # Get total count
    count_query = select(func.count()).select_from(query.subquery())
    result = await db.execute(count_query)
    total = result.scalar_one()

    # Apply pagination and ordering (newest first)
    query = query.offset((page - 1) * page_size).limit(page_size)
    query = query.order_by(ActivityModel.created_at.desc())

    # Execute query
    result = await db.execute(query)
    activities = result.scalars().all()

    return PaginatedActivities(
        data=[Activity.model_validate(a).model_dump() for a in activities],
        total=total,
        page=page,
        page_size=page_size,
        total_pages=(total + page_size - 1) // page_size,
    )


@router.get("/{activity_id}", response_model=ActivityWithRelations)
async def get_activity(
    activity_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Get a single activity by ID with related entity names (single query with joins)."""
    query = (
        select(ActivityModel)
        .where(ActivityModel.id == activity_id)
        .options(
            joinedload(ActivityModel.customer),
            joinedload(ActivityModel.contact),
            joinedload(ActivityModel.order),
            joinedload(ActivityModel.quote),
        )
    )
    result = await db.execute(query)
    activity = result.unique().scalar_one_or_none()

    if not activity:
        raise HTTPException(status_code=404, detail="Activity not found")

    # Build response from eagerly loaded relationships (single query, no N+1)
    activity_data = Activity.model_validate(activity).model_dump()
    activity_data["customer_name"] = activity.customer.company_name if activity.customer else None
    activity_data["contact_name"] = (
        f"{activity.contact.first_name} {activity.contact.last_name}" if activity.contact else None
    )
    activity_data["order_number"] = activity.order.order_number if activity.order else None
    activity_data["quote_number"] = activity.quote.quote_number if activity.quote else None

    return ActivityWithRelations(**activity_data)


@router.post("", response_model=Activity, status_code=201)
async def create_activity(
    activity_data: ActivityCreate,
    db: AsyncSession = Depends(get_db),
):
    """Log a new activity."""
    # Validate related entities if provided
    if activity_data.customer_id:
        customer_query = select(CustomerModel).where(CustomerModel.id == activity_data.customer_id)
        result = await db.execute(customer_query)
        if not result.scalar_one_or_none():
            raise HTTPException(status_code=400, detail="Customer not found")

    if activity_data.contact_id:
        contact_query = select(ContactModel).where(ContactModel.id == activity_data.contact_id)
        result = await db.execute(contact_query)
        if not result.scalar_one_or_none():
            raise HTTPException(status_code=400, detail="Contact not found")

    if activity_data.order_id:
        order_query = select(OrderModel).where(OrderModel.id == activity_data.order_id)
        result = await db.execute(order_query)
        if not result.scalar_one_or_none():
            raise HTTPException(status_code=400, detail="Order not found")

    if activity_data.quote_id:
        quote_query = select(QuoteModel).where(QuoteModel.id == activity_data.quote_id)
        result = await db.execute(quote_query)
        if not result.scalar_one_or_none():
            raise HTTPException(status_code=400, detail="Quote not found")

    # Create activity
    activity = ActivityModel(**activity_data.model_dump())
    db.add(activity)
    await db.commit()
    await db.refresh(activity)

    # Invalidate caches
    await invalidate_cache("api_activities_list")
    await invalidate_cache("activities")

    # Publish real-time events
    await sse_service.publish("dashboard-activity", {
        "activity_type": "crm_activity_logged",
        "title": f"New {activity.activity_type.title()}",
        "description": activity.subject,
        "link": f"/activities/{activity.id}",
        "timestamp": datetime.now(UTC).isoformat(),
    })

    return Activity.model_validate(activity)


@router.put("/{activity_id}", response_model=Activity)
async def update_activity(
    activity_id: UUID,
    activity_data: ActivityUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update an activity."""
    query = select(ActivityModel).where(ActivityModel.id == activity_id)
    result = await db.execute(query)
    activity = result.scalar_one_or_none()

    if not activity:
        raise HTTPException(status_code=404, detail="Activity not found")

    # Update fields
    update_data = activity_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        if field == "activity_type" and value:
            value = value.value  # Convert enum to string
        setattr(activity, field, value)

    await db.commit()
    await db.refresh(activity)

    # Invalidate caches
    await invalidate_cache("api_activities_list")
    await invalidate_cache("activities")

    return Activity.model_validate(activity)


@router.delete("/{activity_id}", status_code=204)
async def delete_activity(
    activity_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Delete an activity."""
    query = select(ActivityModel).where(ActivityModel.id == activity_id)
    result = await db.execute(query)
    activity = result.scalar_one_or_none()

    if not activity:
        raise HTTPException(status_code=404, detail="Activity not found")

    await db.delete(activity)
    await db.commit()

    # Invalidate caches
    await invalidate_cache("api_activities_list")
    await invalidate_cache("activities")

    return None


@router.post("/{activity_id}/complete", response_model=Activity)
async def complete_activity(
    activity_id: UUID,
    complete_data: ActivityComplete | None = None,
    db: AsyncSession = Depends(get_db),
):
    """Mark an activity as complete."""
    query = select(ActivityModel).where(ActivityModel.id == activity_id)
    result = await db.execute(query)
    activity = result.scalar_one_or_none()

    if not activity:
        raise HTTPException(status_code=404, detail="Activity not found")

    # Set completion time
    if complete_data and complete_data.completed_at:
        activity.completed_at = complete_data.completed_at
    else:
        activity.completed_at = datetime.now(UTC)

    await db.commit()
    await db.refresh(activity)

    # Invalidate caches
    await invalidate_cache("api_activities_list")
    await invalidate_cache("activities")

    # Publish real-time event
    await sse_service.publish("dashboard-activity", {
        "activity_type": "activity_completed",
        "title": "Activity Completed",
        "description": activity.subject,
        "timestamp": datetime.now(UTC).isoformat(),
    })

    return Activity.model_validate(activity)


# Entity-specific activity endpoints
@router.get("/customer/{customer_id}", response_model=list[Activity])
async def get_customer_activities(
    customer_id: UUID,
    limit: int = Query(20, ge=1, le=100),
    activity_type: ActivityType | None = None,
    db: AsyncSession = Depends(get_db),
):
    """Get activities for a specific customer (timeline view)."""
    query = select(ActivityModel).where(ActivityModel.customer_id == customer_id)

    if activity_type:
        query = query.where(ActivityModel.activity_type == activity_type.value)

    query = query.order_by(ActivityModel.created_at.desc()).limit(limit)

    result = await db.execute(query)
    activities = result.scalars().all()

    return [Activity.model_validate(a).model_dump() for a in activities]


@router.get("/contact/{contact_id}", response_model=list[Activity])
async def get_contact_activities(
    contact_id: UUID,
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    """Get activities for a specific contact (timeline view)."""
    query = (
        select(ActivityModel)
        .where(ActivityModel.contact_id == contact_id)
        .order_by(ActivityModel.created_at.desc())
        .limit(limit)
    )

    result = await db.execute(query)
    activities = result.scalars().all()

    return [Activity.model_validate(a).model_dump() for a in activities]


@router.get("/pending-tasks", response_model=list[Activity])
async def get_pending_tasks(
    limit: int = Query(10, ge=1, le=50),
    include_overdue: bool = True,
    db: AsyncSession = Depends(get_db),
):
    """Get pending tasks (activities with due dates that are not completed)."""
    query = (
        select(ActivityModel)
        .where(ActivityModel.activity_type == ActivityType.TASK.value)
        .where(ActivityModel.completed_at.is_(None))
        .where(ActivityModel.due_date.isnot(None))
    )

    if not include_overdue:
        query = query.where(ActivityModel.due_date >= datetime.now(UTC))

    query = query.order_by(ActivityModel.due_date.asc()).limit(limit)

    result = await db.execute(query)
    activities = result.scalars().all()

    return [Activity.model_validate(a).model_dump() for a in activities]

File: src/lib/api/activities.ts
/**
 * Activities API Client
 *
 * Provides methods for interacting with the CRM activities endpoints.
 * Activities track customer interactions (calls, emails, meetings, notes, tasks).
 */

import { apiClient } from "./client";
import type {
  Activity,
  ActivityWithRelations,
  CreateActivityRequest,
  UpdateActivityRequest,
  PaginatedActivities,
  ActivityStats,
  ActivityType,
} from "@/types/activities";

/**
 * Parameters for listing activities
 */
export interface ActivityListParams {
  /** Page number (1-indexed) */
  page?: number;
  /** Items per page (default: 50, max: 100) */
  page_size?: number;
  /** Filter by activity type */
  activity_type?: ActivityType;
  /** Filter by customer ID */
  customer_id?: string;
  /** Filter by contact ID */
  contact_id?: string;
  /** Filter by order ID */
  order_id?: string;
  /** Filter by quote ID */
  quote_id?: string;
  /** Include completed tasks in results (default: true) */
  include_completed?: boolean;
}

/**
 * Activities API methods
 */
export const activitiesApi = {
  /**
   * List activities with pagination and filters
   *
   * @param params - Query parameters for filtering and pagination
   * @returns Paginated list of activities with related entity names
   *
   * @example
   * ```ts
   * // Get all activities
   * const activities = await activitiesApi.list();
   *
   * // Get only calls for a specific customer
   * const calls = await activitiesApi.list({
   *   activity_type: ActivityType.CALL,
   *   customer_id: 'customer-123'
   * });
   *
   * // Get pending tasks only
   * const tasks = await activitiesApi.list({
   *   activity_type: ActivityType.TASK,
   *   include_completed: false
   * });
   * ```
   */
  async list(params: ActivityListParams = {}): Promise<PaginatedActivities> {
    const queryParams = new URLSearchParams();

    if (params.page) queryParams.append("page", params.page.toString());
    if (params.page_size) queryParams.append("page_size", params.page_size.toString());
    if (params.activity_type) queryParams.append("activity_type", params.activity_type);
    if (params.customer_id) queryParams.append("customer_id", params.customer_id);
    if (params.contact_id) queryParams.append("contact_id", params.contact_id);
    if (params.order_id) queryParams.append("order_id", params.order_id);
    if (params.quote_id) queryParams.append("quote_id", params.quote_id);
    if (params.include_completed !== undefined) {
      queryParams.append("include_completed", params.include_completed.toString());
    }

    const url = `/api/activities${queryParams.toString() ? `?${queryParams}` : ""}`;
    return apiClient.get<PaginatedActivities>(url);
  },

  /**
   * Get a single activity by ID
   *
   * @param id - Activity UUID
   * @returns Activity with related entity names (customer_name, contact_name, etc.)
   *
   * @example
   * ```ts
   * const activity = await activitiesApi.get('activity-uuid');
   * console.log(activity.customer_name); // "Acme Corp"
   * ```
   */
  async get(id: string): Promise<ActivityWithRelations> {
    return apiClient.get<ActivityWithRelations>(`/api/activities/${id}`);
  },

  /**
   * Create a new activity
   *
   * @param data - Activity creation data
   * @returns Created activity
   *
   * @example
   * ```ts
   * // Log a phone call
   * const call = await activitiesApi.create({
   *   activity_type: ActivityType.CALL,
   *   subject: "Discussed pricing",
   *   description: "Customer interested in bulk pricing",
   *   customer_id: 'customer-123',
   *   contact_id: 'contact-456'
   * });
   *
   * // Create a task with due date
   * const task = await activitiesApi.create({
   *   activity_type: ActivityType.TASK,
   *   subject: "Send quote",
   *   due_date: new Date('2026-02-15').toISOString(),
   *   customer_id: 'customer-123'
   * });
   * ```
   */
  async create(data: CreateActivityRequest): Promise<Activity> {
    return apiClient.post<Activity>("/api/activities", data);
  },

  /**
   * Update an existing activity
   *
   * @param id - Activity UUID
   * @param data - Partial activity update data
   * @returns Updated activity
   *
   * @example
   * ```ts
   * // Update subject
   * await activitiesApi.update('activity-uuid', {
   *   subject: "Updated subject"
   * });
   *
   * // Reschedule task
   * await activitiesApi.update('task-uuid', {
   *   due_date: new Date('2026-02-20').toISOString()
   * });
   * ```
   */
  async update(id: string, data: UpdateActivityRequest): Promise<Activity> {
    return apiClient.put<Activity>(`/api/activities/${id}`, data);
  },

  /**
   * Delete an activity
   *
   * @param id - Activity UUID
   *
   * @example
   * ```ts
   * await activitiesApi.delete('activity-uuid');
   * ```
   */
  async delete(id: string): Promise<void> {
    return apiClient.delete(`/api/activities/${id}`);
  },

  /**
   * Mark a task as complete
   *
   * Sets completed_at timestamp to current time.
   *
   * @param id - Activity UUID (must be a task)
   * @returns Updated activity with completed_at set
   *
   * @example
   * ```ts
   * const completedTask = await activitiesApi.complete('task-uuid');
   * console.log(completedTask.completed_at); // "2026-02-11T10:30:00Z"
   * ```
   */
  async complete(id: string): Promise<Activity> {
    return apiClient.post<Activity>(`/api/activities/${id}/complete`, {});
  },

  /**
   * Get activity timeline for a customer
   *
   * Returns activities chronologically ordered (newest first).
   *
   * @param customerId - Customer UUID
   * @param limit - Maximum activities to return (default: 20)
   * @returns Array of activities for the customer
   *
   * @example
   * ```ts
   * // Get recent customer interactions
   * const timeline = await activitiesApi.getCustomerTimeline('customer-uuid');
   *
   * // Get last 50 interactions
   * const extended = await activitiesApi.getCustomerTimeline('customer-uuid', 50);
   * ```
   */
  async getCustomerTimeline(
    customerId: string,
    limit = 20
  ): Promise<ActivityWithRelations[]> {
    return apiClient.get<ActivityWithRelations[]>(
      `/api/activities/customer/${customerId}?limit=${limit}`
    );
  },

  /**
   * Get activity timeline for a contact
   *
   * Returns activities chronologically ordered (newest first).
   *
   * @param contactId - Contact UUID
   * @param limit - Maximum activities to return (default: 20)
   * @returns Array of activities for the contact
   *
   * @example
   * ```ts
   * const timeline = await activitiesApi.getContactTimeline('contact-uuid');
   * ```
   */
  async getContactTimeline(
    contactId: string,
    limit = 20
  ): Promise<ActivityWithRelations[]> {
    return apiClient.get<ActivityWithRelations[]>(
      `/api/activities/contact/${contactId}?limit=${limit}`
    );
  },

  /**
   * Get pending tasks (incomplete with due dates)
   *
   * Returns tasks that have due dates but are not completed,
   * ordered by due date (soonest first).
   *
   * @returns Array of pending tasks
   *
   * @example
   * ```ts
   * const pendingTasks = await activitiesApi.getPendingTasks();
   *
   * // Check for overdue tasks
   * const overdue = pendingTasks.filter(task => {
   *   return new Date(task.due_date!) < new Date();
   * });
   * ```
   */
  async getPendingTasks(): Promise<ActivityWithRelations[]> {
    return apiClient.get<ActivityWithRelations[]>("/api/activities/pending-tasks");
  },

  /**
   * Get activity statistics
   *
   * Returns aggregated stats about activities:
   * - Count by type (calls, emails, meetings, notes, tasks)
   * - Pending tasks count
   * - Overdue tasks count
   * - Activities completed this week
   *
   * @returns Activity statistics
   *
   * @example
   * ```ts
   * const stats = await activitiesApi.getStats();
   * console.log(stats.by_type[ActivityType.CALL]); // 45
   * console.log(stats.overdue_tasks); // 3
   * ```
   */
  async getStats(): Promise<ActivityStats> {
    return apiClient.get<ActivityStats>("/api/activities/stats");
  },
};
